File: Dynamic Programming/1035. Uncrossed Lines.cpp
class Solution {
    vector<vector<int>> dp;
public:
    int dfs(vector<int>& nums1, vector<int>& nums2, int i1, int i2) {
        if (i1 == nums1.size() || i2 == nums2.size()) return 0;
        // if ()
        if (dp[i1][i2] != -1) return dp[i1][i2];

        int best = 0;
        if (nums1[i1] == nums2[i2]) {
            best = max(best, dfs(nums1, nums2, i1 + 1, i2 + 1) + 1);
        } else {
            best = max(best, dfs(nums1, nums2, i1 + 1, i2));
            best = max(best, dfs(nums1, nums2, i1, i2 + 1));
        }

        dp[i1][i2] = best;
        return best;
    }
    int maxUncrossedLines(vector<int>& nums1, vector<int>& nums2) {
        // when drawing a line, we should always connect to the one which occupies the least width
        // once we connect the numbers, we can no longer use any number prior to them.
        int m = nums1.size();
        int n = nums2.size();
        dp.resize(m, vector<int>(n, -1));

        return dfs(nums1, nums2, 0, 0);


    }


};

File: Dynamic Programming/1035. Uncrossed Lines.py
class Solution:
    def maxUncrossedLines(self, nums1: List[int], nums2: List[int]) -> int:

        m = len(nums1)
        n = len(nums2)
        @cache
        def dfs(i1, i2):
            if i1 == m or i2 == n:
                return 0
            best = 0
            # how do i decide how to connect a number ?
            #
            if nums1[i1] == nums2[i2]:
                best = max(best, dfs(i1 + 1, i2 + 1) + 1)
            else:
                best = max(best, dfs(i1 + 1, i2), dfs(i1, i2 + 1))

            return best

        return dfs(0, 0)